File: java/DivideTwoIntegers.java
// 29. Divide Two Integers
// 29ms

class Solution {
    public int divide(int dividend, int divisor) {
        boolean isNegative = false;
        long dend, dsor, quot, minusor, increment;
        
        if((dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0))
            isNegative = true;
        
        dend = (long)Math.abs((long)dividend);
        dsor = (long)Math.abs((long)divisor);
        
        if(dend < dsor)
            return 0;
        
        if(dsor == 0L)
            return Integer.MAX_VALUE;
        
        minusor = dsor;
        quot = 0L;
        increment = 1L;
        while(dend >= dsor) {
            quot += increment;
            dend -= minusor;
            
            minusor += minusor;
            increment += increment;
            if(minusor > dend) {
                minusor = dsor;
                increment = 1L;
            }
        }
        
        if(isNegative)
            quot = -quot;
        
        if(quot < Integer.MIN_VALUE || quot > Integer.MAX_VALUE)
            quot = (long)Integer.MAX_VALUE;
        
        return (int)quot;
    }
}

File: src/problems/029. Divide Two Integers/DivideTwoIntegers.py
# 29. Divide Two Integers
# Accepted 48ms

class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        if divisor == 0:
            return 2147483647
        
        isNeg = False if (dividend < 0 and divisor < 0) or (dividend >= 0 and divisor >= 0) else True
        dividend, divisor = abs(dividend), abs(divisor)
        
        remainder = dividend
        quotient = 0
        minusor, increment = divisor, 1
        
        while remainder >= divisor:
            if minusor > remainder:
                minusor, increment = divisor, 1
            
            remainder -= minusor
            quotient += increment
            
            minusor += minusor
            increment += increment
        
        quotient = -quotient if isNeg else quotient
        quotient = 2147483647 if quotient < -2147483648 or quotient > 2147483647 else quotient
        
        return quotient
